### PR TITLE
buildroot: optee_os services built as buildroot external package

### PR DESCRIPTION
**Pull request by etienne-lms**
Define external package optee_os_services to build and install
services TAs from optee_os.

The component requires from the buildroot configuretion:
- the cross compile toolchain;
- the path of the optee_os soure file tree;
- the pat of the prebuilt TA devkit.

### CODE REVIEW

File: br-ext/package/optee_os_services/Config.in [hunk 1]
```diff
@@ -0,0 +1,26 @@
+config BR2_PACKAGE_OPTEE_OS_SERVICES
+	bool "optee_os as package to build service TAs installed in fs"
+	select BR2_PACKAGE_OPENSSL
```
Comment (jenswi-linaro) [inline br-ext/package/optee_os_services/Config.in]: Why is OpenSSL selected here?

File: br-ext/package/optee_os_services/optee_os_services.mk [hunk 2]
```diff
@@ -0,0 +1,38 @@
+# This recipe build the services TAs from optee_os and installs them in the
+# root file system at path /lib/optee_armtz/.
+#
+# The path of optee_os source file tree is expected from BR2_PACKAGE_OPTEE_OS_SERVICES_SITE.
+# The path of the TA devkit is expected from BR2_PACKAGE_OPTEE_OS_SERVICES_SDK.
+# The cross compiler is expected from BR2_PACKAGE_OPTEE_OS_SERVICES_CROSS_COMPILE (Alternative: could use TARGET_COMPILE).
```
Comment (jenswi-linaro) [inline br-ext/package/optee_os_services/optee_os_services.mk]: This line seem a little long

Comment (etienne-lms) [reply br-ext/package/optee_os_services/optee_os_services.mk]: Prefix `BR2_PACKAGE_OPTEE_OS_SERVICES_` is mandatory.

For the compiler, can't we simply rely on `TARGET_CROSS` in the recipe, and not provide a `BR2_xxx`?

File: br-ext/package/optee_os_services/Config.in (hunk 1 above)
Comment (etienne-lms) [reply br-ext/package/optee_os_services/Config.in]: true. will remove.

File: br-ext/package/optee_os_services/optee_os_services.mk (hunk 2 above)
Comment (jenswi-linaro) [reply br-ext/package/optee_os_services/optee_os_services.mk]: The name of the variable is fine, it's just that the line is 122 characters, where 80 is max.

Comment (etienne-lms) [reply br-ext/package/optee_os_services/optee_os_services.mk]: ok, sorry.

File: br-ext/package/optee_os_services/Config.in [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+if BR2_PACKAGE_OPTEE_OS_SERVICES
+
+config BR2_PACKAGE_OPTEE_OS_SERVICES_SITE
+	string "OP-TEE OS path"
+	default ""
+	help
+	  The path to the source directory of OP-TEE test
+
+config BR2_PACKAGE_OPTEE_OS_SERVICES_SDK
+	string "OPTEE SDK path"
```
Comment (jbech-linaro) [inline br-ext/package/optee_os_services/Config.in]: Some nits: When it is **not** configuration flags, then please 
s/OPTEE/OP-TEE/

File: br-ext/package/optee_os_services/Config.in [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	help
+	  TA devkit to use for building service TAs from optee-os.
+
+config BR2_PACKAGE_OPTEE_OS_SERVICES_CROSS_COMPILE
+	string "OPTEE cross compiler"
```
Comment (jbech-linaro) [inline br-ext/package/optee_os_services/Config.in]: s/OPTEE/OP-TEE/

Comment (etienne-lms) [reply br-ext/package/optee_os_services/Config.in]: done

File: br-ext/package/optee_os_services/Config.in (hunk 3 above)
Comment (etienne-lms) [reply br-ext/package/optee_os_services/Config.in]: done

**Comment by etienne-lms** [comment]
comments addressed.
@jenswi-linaro tag applied. 
rebased.

File: br-ext/package/optee_os_services/Config.in [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+config BR2_PACKAGE_OPTEE_OS_SERVICES_SITE
+	string "OP-TEE OS path"
+	default ""
+	help
+	  The path to the source directory of OP-TEE test
+
+config BR2_PACKAGE_OPTEE_OS_SERVICES_SDK
+	string "OP-TEE SDK path"
+	help
+	  TA devkit to use for building service TAs from optee-os.
```
Comment (jbech-linaro) [inline br-ext/package/optee_os_services/Config.in]: s/op-tee/OP-TEE/

File: br-ext/package/optee_os_services/optee_os_services.mk [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+OPTEE_OS_SERVICES_SOURCE = local
+OPTEE_OS_SERVICES_SITE = $(BR2_PACKAGE_OPTEE_OS_SERVICES_SITE)
+OPTEE_OS_SERVICES_SITE_METHOD = local
+OPTEE_OS_SERVICES_INSTALL_STAGING = YES
+
+OPTEE_OS_SERVICES_LICENSE = BSD-2-Clause
+OPTEE_OS_SERVICES_LICENSE_FILES = LICENSE
+
+define OPTEE_OS_SERVICES_BUILD_CMDS
+	@$(foreach f,$(wildcard $(@D)/ta_services/*/Makefile), \
```
Comment (jbech-linaro) [inline br-ext/package/optee_os_services/optee_os_services.mk]: I think  the last we said wasn't to use `ta_services` and instead just use `ta`. I.e., this should be updated accordingly.

File: br-ext/package/optee_os_services/optee_os_services.mk [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+		echo Building $f && \
+			$(MAKE) CROSS_COMPILE="$(shell echo $(BR2_PACKAGE_OPTEE_OS_SERVICES_CROSS_COMPILE))" \
+			O=out TA_DEV_KIT_DIR=$(call qstrip,$(BR2_PACKAGE_OPTEE_OS_SERVICES_SDK)) \
+			$(TARGET_CONFIGURE_OPTS) -C $(dir $f) all &&) true
+endef
+
+# May installs OP-TEE TA services in $(TARGET_DIR)/lib/optee_armtz
```
Comment (jbech-linaro) [inline br-ext/package/optee_os_services/optee_os_services.mk]: s/installs/install/

File: br-ext/package/optee_os_services/optee_os_services.mk (hunk 6 above)
Comment (etienne-lms) [reply br-ext/package/optee_os_services/optee_os_services.mk]: ok, fine. let's move service TAs into optee_os/ta/...

**Comment by etienne-lms** [comment]
Rebased, comment addressed, removed tag, waiting for ack that change is ok to squash and tag back.

**Comment by jenswi-linaro** [comment]
I'm starting to wonder if this is the right way to do it. In https://github.com/jenswi-linaro/optee_os/blob/keymaster/ta_services/ta_services.mk (ironically using ta_services instead) we're building TAs from optee_os git directly.

For the SKS TA building in a separate step as in this PR will work. However, if we have a TA we'd like to include as an "early TA" it will not work any longer so at some point we'll need to be able to build TAs while building OP-TEE OS and Dev-kit.

Do we care about that right now? I'd like to hear the opinion of others on this.

**Comment by jforissier** [comment]
@jenswi-linaro I also think we should be able to build at the same time as optee_os and dev kit. "Early TA" is a good justification for this, indeed.

What you did in the keymaster prototype fits the bill, although I would like to see it done differently if possible. That is, do not invoke make recursively and have rules for each individual object file etc. instead (like we do for all user-mode libraries basically). This would probably require some refactoring of the dev kit files to avoid duplicating flags. Not exactly a funny thing to do admittedly, so the recursive build might be OK as a first step...

**Comment by etienne-lms** [comment]
I think optee_os should support both, allowing one to build the whole optee_os (core/libs/TAs) or to build libs and/or TAs from an already built (delivered) core.

**Comment by jenswi-linaro** [comment]
I see how it can be useful to build a TA after delivery (reconfiguration etc).

Let's do both. I'll take care of the building of TA at the same time as optee_os, if no one beats me to it... :-)

I'm still OK with this PR, please squash etc.

**Comment by etienne-lms** [comment]
Squashed, tag applied.

**Pull request closed without merging**